Add Windows support for system_info probe

jan-cerny opened the pull request:
The system_info probe retreives operating system name, operating
system version, names of network interfaces, their MAC and IP
addresses. On Windows, we can get these information using Windows
API calls and then populate them into OVAL item.

See the example OVAL Results Document.
[results.zip](https://github.com/OpenSCAP/openscap/files/1853105/results.zip)

mpreisler commented on src/OVAL/probes/independent/system_info.c:
```diff
@@ -535,7 +681,35 @@ static char * _offline_chroot_get_hname(void)
+	RtlGetVersion(&osverinfo);
+
+	char *version = malloc(VERSION_LEN);
+	snprintf(version, VERSION_LEN, "%d.%d.%d", osverinfo.dwMajorVersion, osverinfo.dwMinorVersion, osverinfo.dwBuildNumber);
```
any reason not to use oscap_sprintf?

jan-cerny replied:
I always forget about existence of this function :smile:

mpreisler commented:
A lot of this code breaks our code conventions, mainly the variable names.

jan-cerny commented:
> A lot of this code breaks our code conventions, mainly the variable names.

How it should look like?

mpreisler commented:
Labeled do not merge becuase this is blocked by another PR.

jan-cerny commented:
I have rebased this on the top of master branch and I have resolved conflicts with the master branch.

mpreisler commented:
@openscap-ci test this please

mpreisler commented:
> How it should look like?

It would be really strenuous to mention every single instance of it breaking our code conventions, so here is one example:
```
PIP_ADAPTER_ADDRESSES pAddresses = NULL;
PIP_ADAPTER_ADDRESSES pCurrAddresses = NULL;
```
We do not use camelcase variable names. It's not really a big deal, I guess if you took the code from other sources cite them properly, make sure licensing works and then we can keep them in their original conventions. If you wrote this from scratch then why are you using WinAPI conventions?

mpreisler commented on src/OVAL/probes/independent/system_info.c:
```diff
@@ -295,6 +305,142 @@ static int get_ifs(SEXP_t *item)
+	do {
+		pAddresses = malloc(WORKING_BUFFER_SIZE);
+		dwRetVal = GetAdaptersAddresses(family, flags, NULL, pAddresses, &outBufLen);
+		if (dwRetVal == ERROR_BUFFER_OVERFLOW) {
```
if this fails once why would it ever work again in any other try? shouldn't you increase buffer size if it fails or something like that?

jan-cerny replied:
Nice catch, thatk you. I think here  should be outBufLen passed to malloc as parameter.

mpreisler commented on src/OVAL/probes/independent/system_info.c:
```diff
@@ -295,6 +305,142 @@ static int get_ifs(SEXP_t *item)
+			const size_t mac_address_str_len = pCurrAddresses->PhysicalAddressLength * 3;
+			mac_address_str = malloc(mac_address_str_len);
+
+			for (int i = 0; i < (int)pCurrAddresses->PhysicalAddressLength;
```
if length is unsigned and i is also unsigned why not just use unsigned instead of casting here?

mpreisler commented on src/common/util.c:
```diff
@@ -338,3 +339,13 @@ char *oscap_path_join(const char *path1, const char *path2)
 }
+
+#ifdef _WIN32
+char *oscap_wstr_to_str(wchar_t *wstr)
```
wstr and wchar have a ton of definitions, I suggest to prefix this with win to make it clear this is windows only. wchar can have different widths depending on the OS and compiler.

jan-cerny replied:
+1

mpreisler commented on src/common/util.c:
```diff
@@ -338,3 +339,13 @@ char *oscap_path_join(const char *path1, const char *path2)
+#ifdef _WIN32
+char *oscap_wstr_to_str(wchar_t *wstr)
+{
+	size_t required_size = WideCharToMultiByte(CP_UTF8, 0, wstr, -1, NULL, 0, NULL, NULL);
```
const

jan-cerny commented:
I see, I should use variable names in our conventions. I usually named them after the function parameters that I have found in docs. I will change that.

I am not sure about the data types, though. Some of the Windows data types are just aliases for C dtata types. For example `DWORD` is defined as `typedef unsigned long DWORD` but I don't know if it is preferred to use `unsigned long` or `DWORD` in the code that interacts with Windows API.

jan-cerny commented:
I have also realized that I should refactor printing the error messages and use there wide character strings there. I will try to change that.

mpreisler commented:
> For example DWORD is defined as typedef unsigned long DWORD but I don't know if it is preferred to use unsigned long or DWORD in the code that interacts with Windows API.

Use Windows typedefs in this case, using a different typedef is not a coding style, that's different code with potentially different behavior on different platform/OS.

mpreisler commented:
@openscap-ci test this please

jan-cerny commented on src/OVAL/probes/independent/system_info_probe.c:
```diff
@@ -565,11 +731,39 @@ int system_info_probe_main(probe_ctx *ctx, void *arg)
+	GetNativeSystemInfo(&si);
+	const char *arch;
+	switch (si.wProcessorArchitecture) {
+	case PROCESSOR_ARCHITECTURE_AMD64:
```
Maybe it would be better to put this out from main function and create something like `static const char *get_windows_architecture(void)`.

jan-cerny replied:
done

jan-cerny commented:
I have improved the code and I have rebased it on master. I have changed variables and functions names, refactored the error messages, and I have improved the code style.

jan-cerny commented:
@openscap-ci test this please

mpreisler commented:
@openscap-ci test this please

jan-cerny commented:
@openscap-ci test this please

mpreisler commented:
ACK
